userdb/memory/mem.go

```diff
@@ -0,0 +1,96 @@
+package memory
+
+import (
+	"bufio"
+	"encoding/csv"
+	"io"
+	"os"
+	"sync"
+
+	"github.com/janekolszak/idp/core"
+	"golang.org/x/crypto/bcrypt"
+)
+
+type Store struct {
+	hashes map[string]string
+	mtx    sync.RWMutex
+}
+
+func NewMemStore() (*Store, error) {
+	s := Store{}
+
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	s.hashes = make(map[string]string)
+
+	return &s, nil
+}
+
+func (s *Store) LoadHtpasswd(filename string) error {
+
+	f, err := os.OpenFile(filename, os.O_RDONLY, os.ModeExclusive)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	r := csv.NewReader(bufio.NewReader(f))
+	r.Comment = '#'
+	r.Comma = ':'
+	r.TrimLeadingSpace = true
+	r.FieldsPerRecord = 2
+
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	s.hashes = make(map[string]string)
+	for {
+		fields, err := r.Read()
+		if err != nil {
+			if err == io.EOF {
+				return nil
+			}
+
+			return err
+		}
+
+		s.hashes[fields[0]] = fields[1]
+	}
+}
+
+func (s *Store) Check(username, password string) error {
```

Comment:
I integrated htpasswd into this `memory` userdb implementation.  It didn't start out like that but then I noticed I'd pretty much replicated the functionality you had in htpasswd so it seemed to make sense to combine.  The userdb implementation is the bit that does know how it's stored .. the interface in `userdb/userdb.go` ... I was looking towards providing either sqlite/etc in future which would serve as alternate implementation of the interface.